### PR TITLE
fix(downtime datepicker): update end time properly when start time or…

### PR DESCRIPTION
**Pull request by miteto**
<h1> Fix downtime datepicker </h1>

<h2> Description </h2>

Fix end time when start time or duration is updated

Fixes 3642

<h2> Type of change </h2>

- [X] Patch fixing an issue (non-breaking change)

<h2> Target serie </h2>

- [ ] 2.8.x
- [X] 18.10.x
- [X] 19.04.x (master)

### CODE REVIEW

**Comment by CPbN** [comment]
Just for ref, anoter downtime duration & end date PR is #7357.

File: www/include/monitoring/downtime/template/AddDowntime.ihtml [hunk 1]
```diff
@@ -154,5 +154,37 @@
 
     initDatepicker("datepicker", "mm/dd/yy", "0");
 
+    function updateEndTime() {
```
Comment (sc979) [inline www/include/monitoring/downtime/template/AddDowntime.ihtml]: As this issue also impact other pages (ie: recurrent) or widgets' popin, you need to move this new method in :
`www/include/common/javascript/datepicker/localizedDatepicker.js` and call it on every pages using an endDate's datepicker

File: www/include/monitoring/downtime/template/AddDowntime.ihtml [hunk 2]
```diff
@@ -154,5 +154,37 @@
...
+            } else if (duration_scale == 'h') {
+                dur = $("#duration").val() * 3600;
+            } else if (duration_scale == 'd') {
+                dur = $("#duration").val() * 86400;
+            }
+            dur *= 1000; // Convert to microseconds
+            var e = new Date();
+            e.setTime(d.getTime()+dur);
+            $(".datepicker").last().datepicker("setDate", e);
+            $(".timepicker").last().timepicker("setTime", e.getHours()+':'+e.getMinutes());
```
Comment (sc979) [inline www/include/monitoring/downtime/template/AddDowntime.ihtml]:
```suggestion
            $(".timepicker").last().timepicker("setTime", e.getHours() + ':' + e.getMinutes());
```

File: www/include/monitoring/downtime/template/AddDowntime.ihtml [hunk 3]
```diff
@@ -154,5 +154,37 @@
...
+            } else if (duration_scale == 'm') {
+                dur = $("#duration").val() * 60;
+            } else if (duration_scale == 'h') {
+                dur = $("#duration").val() * 3600;
+            } else if (duration_scale == 'd') {
+                dur = $("#duration").val() * 86400;
+            }
+            dur *= 1000; // Convert to microseconds
+            var e = new Date();
+            e.setTime(d.getTime()+dur);
```
Comment (sc979) [inline www/include/monitoring/downtime/template/AddDowntime.ihtml]:
```suggestion
            e.setTime(d.getTime() + dur);
```

File: www/include/monitoring/downtime/template/AddDowntime.ihtml [hunk 4]
```diff
@@ -154,5 +154,37 @@
...
+        }
+    }
+
+    $(".datepicker").first().on('change', function(e) {
+        updateEndTime();
+    });
+    $("#duration").on('keypress', function(e) {
+        updateEndTime();
+    });
+    $("#duration_scale").on('change', function(e) {
```
Comment (sc979) [inline www/include/monitoring/downtime/template/AddDowntime.ihtml]: only a comment :
are you sure that when a duration is chosen, we need to second datepicker to be set ?
isn't to the BE to convert the duration in a timestamp ?

Comment (miteto) [reply www/include/monitoring/downtime/template/AddDowntime.ihtml]: We are updating end date based on start date and duration. When duration changes, end date is updated as well. Of course, user can overwrite end date datepicker (manually updating it), but I think that, when, for example duration is changed from 1 to 2 days, end date should be updated accordingly.

File: www/include/common/javascript/datepicker/localizedDatepicker.js [hunk 5]
```diff
@@ -113,3 +113,36 @@ function setUserFormat() {
...
+            dur = $("#duration").val() * 86400;
+        }
+        dur *= 1000; // Convert to microseconds
+        var e = new Date();
+        e.setTime(d.getTime() + dur);
+        $(".datepicker").last().datepicker("setDate", e);
+        $(".timepicker").last().timepicker("setTime", e.getHours() + ':' + e.getMinutes());
+    }
+}
+$(document).ready(function() {
```
Comment (sc979) [inline www/include/common/javascript/datepicker/localizedDatepicker.js]: shouldn't this part of the code be called in the page ?

**Comment by sc979** [comment]
I approved the PR as it seems to correct the datepicker issue in this page. But you should add these new methods to the services templates and widgets' popin template too ;)

**Pull request closed without merging**